lib/hdk_time_indexing/src/writing.rs

```diff
@@ -0,0 +1,97 @@
+{
+    // write the time index tree
+    let leafmost_segment = ensure_time_index(index_name, time)?;
+    let leafmost_hash = leafmost_segment.hash()?;
+
+    // create a virtual segment for determining the final link tag data
+    let target_entry_segment = IndexSegment::leafmost_link(&time);
+    let encoded_link_tag = target_entry_segment.tag_for_index(&index_name);
+
+    // ensure link from the leaf index to the target entry
+    link_if_not_linked(leafmost_hash.to_owned(), entry_hash.to_owned(), encoded_link_tag.to_owned())?;
+
+    // ensure a reciprocal link from the target entry back to the leaf index node
+    link_if_not_linked(entry_hash, leafmost_hash, encoded_link_tag)?;
+
+    Ok(())
+}
+
+/// Returns the leaf-most `IndexSegment` in the time tree, so that target entries can be
+/// linked from it.
+///
+fn ensure_time_index<I>(index_name: &I, time: DateTime<Utc>) -> TimeIndexResult<IndexSegment>
+    where I: AsRef<str>,
+{
+    // create a root anchor for the path based on the index name
+    let root = Path::from(index_name.as_ref());
+    root.ensure()?;
+    let root_hash = root.path_entry_hash()?;
+
+    let segments = get_index_segments(&time);
+
+    for (idx, segment) in segments.iter().enumerate() {
+        segment.ensure()?;
+
+        if idx == 0 {
+            // link the first segment to the root
+            if !segment_links_exist(index_name, &root_hash, segment)? {
+                create_link(
+                    root_hash.to_owned(),
+                    segment.hash()?,
+                    HdkLinkType::Any,
+                    segment.tag_for_index(&index_name),
+                )?;
+            }
+        } else {
+            // link subsequent segments to the previous one
+            let prev_segment_hash = segments.get(idx - 1).unwrap().hash()?;
+
+            if !segment_links_exist(index_name, &prev_segment_hash, segment)? {
+                create_link(
+                    prev_segment_hash,
+                    segment.hash()?,
+                    HdkLinkType::Any,
+                    segment.tag_for_index(&index_name),
+                )?;
+            }
+        }
+    }
+
+    Ok(segments.last().unwrap().cloned())
+}
+
+fn segment_links_exist<I>(index_name: &I, base_hash: &EntryHash, target_segment: &IndexSegment) -> TimeIndexResult<bool>
+    where I: AsRef<str>,
+{
+    Ok(get_links(base_hash.to_owned(), Some(target_segment.tag_for_index(&index_name)))?
+        .len() > 0)
+}
+
+fn link_if_not_linked(origin_hash: EntryHash, dest_hash: EntryHash, link_tag: LinkTag) -> TimeIndexResult<()> {
+    if false == get_links(origin_hash.to_owned(), Some(link_tag.to_owned()))?
+        .iter().any(|l| { EntryHash::from(l.target.to_owned()) == dest_hash })
+    {
+        create_link(
+            dest_hash.to_owned(),
+            origin_hash.to_owned(),
+            HdkLinkType::Any,
+            link_tag,
+        ).map_err(|_e| { TimeIndexingError::EntryNotFound(origin_hash.to_owned()) })?;
```

Comment:
you mean the error mapping? (sorry, it's not clear exactly which line you mean in the mobile Github UI :S)

Reply:
The error mapping yeah

Reply:
Gotcha. Well, "real" behaviour- probably not?
I'm moving to a pattern where records have to be committed to the local DHT before any indexes can be created for them. Calls to `create_link` fail if the entry being linked to doesn't exist, and I wanted to catch this as the library's own error type so that the reason for erroring is a bit clearer to the caller. If you check `append_to_time_index` you'll see that it ensures the pointer entry being linked to exists before populating the index data structures.